File: core/src/dc/targetman/ai/Segment.java
package dc.targetman.ai;

import java.util.HashSet;
import java.util.Set;

import com.badlogic.gdx.math.Rectangle;

import dclib.geometry.RectangleUtils;
import dclib.util.Maths;

class Segment {

	Rectangle bounds;
	DefaultNode leftNode;
	DefaultNode rightNode;
	Set<DefaultNode> nodes = new HashSet<DefaultNode>();

	Segment(final Rectangle bounds) {
		this.bounds = bounds;
		leftNode = new DefaultNode(bounds.x, y());
		nodes.add(leftNode);
		rightNode = new DefaultNode(RectangleUtils.right(bounds), y());
		nodes.add(rightNode);
	}
	
	float left() {
		return leftNode.x();
	}
	
	float right() {
		return rightNode.x();
	}
	
	float y() {
		return RectangleUtils.top(bounds);
	}

	boolean containsX(final float x) {
		return Maths.between(x, left(), right());
	}

	boolean overlapsX(final Rectangle bounds) {
		float boundsRight = RectangleUtils.right(bounds);
		return Maths.between(left(), bounds.x, boundsRight)
				|| Maths.between(right(), bounds.x, boundsRight)
				|| Maths.between(bounds.x, left(), right())
				|| Maths.between(boundsRight, left(), right());
	}

}

File: core/src/main/dc/targetman/ai/graph/Segment.kt
package dc.targetman.ai.graph

import com.badlogic.gdx.math.Rectangle
import dclib.geometry.right
import dclib.geometry.top

class Segment(val bounds: Rectangle) {
    val leftNode = DefaultNode(bounds.x, y)
    val rightNode = DefaultNode(bounds.right, y)
    val left get() = leftNode.x
    val right get() = rightNode.x
    val y get() = bounds.top

    private val nodes: MutableSet<DefaultNode>

    init {
        nodes = mutableSetOf(leftNode, rightNode)
    }

    fun getNodes(): Set<DefaultNode> {
        return nodes.toSet()
    }

    fun getOrCreateNode(nodeX: Float): DefaultNode {
        return nodes.firstOrNull { it.x == nodeX } ?: createNode(nodeX)
    }

    fun createNode(nodeX: Float): DefaultNode {
        val node = DefaultNode(nodeX, y)
        nodes.add(node)
        return node
    }

    fun containsX(x: Float): Boolean {
        return x in left..right
    }

    fun overlapsX(bounds: Rectangle): Boolean {
        return left in bounds.x..bounds.right
                || right in bounds.x..bounds.right
                || bounds.x in left..right
                || bounds.right in left..right
    }

    override fun toString(): String {
        return bounds.toString()
    }
}
